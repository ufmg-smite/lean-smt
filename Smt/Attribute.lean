import Lean
import Std

namespace Smt.Attribute

open Lean
open Std

/-- An extension to Lean's runtime environment to support SMT attributes.
    Maintains a set of function declarations for the `smt` tactic to utilize
    while generating the SMT query. -/
abbrev SmtExtension := SimpleScopedEnvExtension Name (HashSet Name)

/-- Adds a declaration to the set of function declarations maintained by the SMT
    environment extension. -/
def addSmtEntry (d : HashSet Name) (e : Name) : HashSet Name :=
  d.insert e

initialize smtExt : SmtExtension ← registerSimpleScopedEnvExtension {
  name     := `SmtExt
  initial  := {}
  addEntry := addSmtEntry
}

/-- Throws unexpected type error. -/
def throwUnexpectedType (t : Name) (n : Name) : AttrM Unit :=
  throwError s!"unexpected type at '{n}', `{t}` expected"

/-- Validates the tagged declaration. -/
def validate (n : Name) : AttrM Unit := do
  match (← getEnv).find? n with
  | none      => throwError s!"unknown constant '{n}'"
  | some info =>
    match info.type with
    | Expr.const c .. =>
      if c != (`Smt.Transformer) then throwUnexpectedType `Smt.Transformer n
    | _               => throwUnexpectedType `Smt.Transformer n

/-- Registers an SMT attribute with the provided name and description and links
    it against `ext`. -/
def registerSmtAttr (attrName : Name) (attrDescr : String)
  : IO Unit :=
  registerBuiltinAttribute {
    name  := attrName
    descr := attrDescr
    applicationTime := AttributeApplicationTime.afterTypeChecking
    add   := fun decl stx attrKind => do
      trace[Smt.debug.attr] s!"attrName: {attrName}, attrDescr: {attrDescr}"
      trace[Smt.debug.attr] s!"decl: {decl}, stx: {stx}, attrKind: {attrKind}"
      Attribute.Builtin.ensureNoArgs stx
      validate decl
      setEnv (smtExt.addEntry (← getEnv) decl)
      trace[Smt.debug.attr]
        s!"transformers: {(smtExt.getState (← getEnv)).toList}"
    erase := fun declName => do
      let s := smtExt.getState (← getEnv)
      let s := s.erase declName
      modifyEnv fun env => smtExt.modifyState env fun _ => s
  }

initialize registerSmtAttr `Smt
             "Utilize this function to transform Lean expressions to SMT terms."

end Smt.Attribute
