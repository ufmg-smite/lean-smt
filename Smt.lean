import Smt.General
import Smt.Nat
import Smt.Prop
import Smt.Solver
import Smt.String
import Smt.Tactic
import Smt.Term
