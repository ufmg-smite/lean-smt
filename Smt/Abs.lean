/-
Copyright (c) 2021-2023 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean
import Smt.Translator
import Mathlib.Algebra.Order.Group.Abs

namespace Smt.Abs

open Lean Expr
open Translator Term


@[smtTranslator] def replaceAbs : Translator
  | app (app (app (app (const ``abs _) _) _) _) e => do
      let tmE ← applyTranslators! e
      return appT (symbolT "abs") tmE
  | _ => return none

end Smt.Abs
