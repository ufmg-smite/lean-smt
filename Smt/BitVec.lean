/-
Copyright (c) 2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Wojciech Nawrocki
-/

import Lean
import Smt.Translator
import Smt.Data.BitVec

namespace Smt.BitVec

open Lean Expr
open Translator Term

@[smtTranslator] def replaceType : Translator
  | e@(app (const ``BitVec _) n) => do
    let n ← Meta.whnf n
    let some n ← Meta.evalNat n |>.run
      | throwError "bitvector type{indentD e}\nhas variable width"
    return mkApp2 (symbolT "_") (symbolT "BitVec") (literalT (toString n))
  | _ => return none

def mkLit (w : Nat) (n : Nat) : Term :=
  let bits := Nat.toDigits 2 n |>.take w
  literalT <| bits.foldl (init := "#b".pushn '0' (w - bits.length)) (·.push ·)

@[smtTranslator] def replaceEq : Translator
  -- TODO: we should really just support beq/bne across all types uniformly
  | app (app (const ``BEq.beq _) (app (const ``BitVec _) _)) _ => return symbolT "="
  | app (app (app (app (const ``bne _) (app (const ``BitVec _) _)) _) e) e' =>
    return appT (symbolT "not") (mkApp2 (symbolT "=") (← applyTranslators! e) (← applyTranslators! e'))
  | _ => return none

@[smtTranslator] def replaceFun : Translator
  | app (app (const ``BitVec.append _) _) _ => return symbolT "concat"
  | app (const ``BitVec.and _) _            => return symbolT "bvand"
  | app (const ``BitVec.or _) _             => return symbolT "bvor"
  | app (const ``BitVec.xor _) _            => return symbolT "bvxor"
  | e@(app (app (app (const ``BitVec.shiftLeft _) w) x) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkApp2 (symbolT "bvshl") (← applyTranslators! x) (mkLit w n)
  | e@(app (app (app (const ``BitVec.shiftRight _) w) x) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkApp2 (symbolT "bvlshr") (← applyTranslators! x) (mkLit w n)
  | e@(app (app (app (const ``BitVec.extract _) _) i) j) => do
    let i ← reduceLit i e
    let j ← reduceLit j e
    return mkApp3 (symbolT "_") (symbolT "extract") (literalT (toString i)) (literalT (toString j))
  | e@(app (const ``BitVec.zero _) w) => do
    let w ← reduceWidth w e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkLit w 0
  | e@(app (app (const ``BitVec.ofNat _) w) n) => do
    let w ← reduceWidth w e
    let n ← reduceLit n e
    if w == 0 then throwError "cannot emit bitvector literal{indentD e}\nof bitwidth 0"
    return mkLit w n
  | _ => return none
where
  reduceWidth (w : Expr) (e : Expr) : TranslationM Nat := do
    let some w ← Meta.evalNat (← Meta.whnf w) |>.run
      | throwError "bitvector width{indentD w}\nis not constant in{indentD e}"
    return w
  reduceLit (n : Expr) (e : Expr) : TranslationM Nat := do
    let some n ← Meta.evalNat (← Meta.whnf n) |>.run
      | throwError "literal{indentD n}\nis not constant in{indentD e}"
    return n

end Smt.BitVec
