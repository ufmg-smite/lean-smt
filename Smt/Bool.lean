/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean
import Smt.Transformer

namespace Smt.Bool

open Lean Expr
open Smt.Transformer

@[Smt] def replaceConst : Transformer
  | const `Bool.true ..                                 => pure (mkConst `true)
  | const `Bool.false ..                                => pure (mkConst `false)
  | app (app (const `BEq.beq ..) (const `Bool ..) _) .. =>
    pure (mkConst (Name.mkSimple "="))
  | e                                                   => pure e

@[Smt] def replaceEq : Transformer
  | app (app (const `Decidable.decide ..)
             (app (app (app (const `Eq ..) (const `Bool ..) _)
                       a _) b _) ..) .. =>
    return match ← applyTransformations a, ← applyTransformations b with
    | some a', some b' =>
      mkApp2 (mkConst (Name.mkSimple "=")) a' b'
    | _      , _       => none
  | e                                   => pure e

end Smt.Bool
