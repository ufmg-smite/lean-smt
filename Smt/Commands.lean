/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed, Tomaz Gomes Mascarenhas, Wojciech Nawrocki
-/

import Smt.Solver
import Smt.Term

namespace Smt

/-- An SMT-LIBv2 command that we can emit. -/
inductive Command where
  | assert (tm : Term)
  | declare (nm : String) (st : Term)
  | declareSort (nm : String) (arity : Nat)
  | defineSort (nm : String) (ps : List Term) (tm : Term)
  | defineFun (nm : String) (ps : List (String × Term)) (cod : Term) (tm : Term) (rec : Bool)

namespace Command

open Lean
open Term
open scoped Term.Notation

def defNat : Command := .defineSort "Nat" [] (`"Int")

def defNatSub : Command :=
  .defineFun "Nat.sub" [("x", `"Nat"), ("y", `"Nat")] (`"Nat")
    (`"ite" • (`"<" • `"x" • `"y") • ``"0" • (`"-" • `"x" • `"y"))
    false

def sortEndsWithNat : Term → Bool
  | arrowT _ t    => sortEndsWithNat t
  | symbolT "Nat" => true
  | _             => false

def natAssertBody (t : Term) : Term :=
  mkApp2 (symbolT ">=") t (literalT "0")

variable [Monad m] [MonadNameGenerator m]

/-- TODO: refactor to support functions as input (e.g., (Nat → Nat) → Nat). -/
def natConstAssert (n : String) (args : List Name) : Term → m Term
  | arrowT i@(symbolT "Nat") t => do
    let id ← Lean.mkFreshId
    return (forallT id.toString i
                   (imp id.toString (← natConstAssert n (id::args) t)))
  | arrowT a t => do
    let id ← Lean.mkFreshId
    return (forallT id.toString a (← natConstAssert n (id::args) t))
  | _ => pure $ natAssertBody (applyList n args)
  where
    imp n t := appT (appT (symbolT "=>") (natAssertBody (symbolT n))) t
    applyList n : List Name → Term
      | [] => symbolT n
      | t :: ts => appT (applyList n ts) (symbolT t.toString)

def emitCommand (s : Solver) (cmd : Command) : m Solver := do
  let mut s := s
  match cmd with
  | .assert val                   => s := s.assert val
  | .declare nm st@(arrowT ..)    => s := s.declareFun nm st
  | .declare nm st                => s := s.declareConst nm st
  | .declareSort nm arity         => s := s.declareSort nm arity
  | .defineSort nm ps tm          => s := s.defineSort nm ps tm
  | .defineFun nm ps cod tm true  => s := s.defineFunRec nm ps cod tm
  | .defineFun nm ps cod tm false => s := s.defineFun nm ps cod tm
  match cmd with
  | .declare nm st =>
    if sortEndsWithNat st then
      s := s.assert (← natConstAssert nm [] st)
  | .defineFun nm ps cod _ _ =>
    if sortEndsWithNat cod then
      let tmArrow := ps.foldr (init := cod) fun (_, tp) acc => arrowT tp acc
      s := s.assert (← natConstAssert nm [] tmArrow)
  | _ => pure ()
  return s

end Command
end Smt
