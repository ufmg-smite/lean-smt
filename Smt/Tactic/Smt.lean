/-
Copyright (c) 2021-2022 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Lean

import Smt.Dsl.Sexp
import Smt.Query
import Smt.Reconstruction.Certifying
import Smt.Solver

namespace Smt

open Lean Elab Tactic
open Smt Query Solver

initialize
  registerTraceClass `smt.debug
  registerTraceClass `smt.debug.attr
  registerTraceClass `smt.debug.reconstruct
  registerTraceClass `smt.debug.translate.query
  registerTraceClass `smt.debug.translate.expr

syntax smtHints := ("[" ident,* "]")?
syntax smtTimeout := ("(timeout := " num ")")?

/-- `smt` converts the current goal into an SMT query and checks if it is
satisfiable. By default, `smt` generates the minimum valid SMT query needed to
assert the current goal. However, that is not always enough:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt
```
For the theorem above, `smt` generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(check-sat)
```
which is missing the hypotheses `hp` and `hpq` required to prove the theorem. To
pass hypotheses to the solver, use `smt [h₁, h₂, ..., hₙ]` syntax:
```lean
def modus_ponens (p q : Prop) (hp : p) (hpq : p → q) : q := by
  smt [hp, hpq]
```
The tactic then generates the query below:
```smt2
(declare-const q Bool)
(assert (not q))
(declare-const p Bool)
(assert p)
(assert (=> p q))
(check-sat)
```
-/
syntax (name := smt) "smt" smtHints smtTimeout : tactic

/-- Like `smt`, but just shows the query without invoking a solver. -/
syntax (name := smtShow) "smt_show" smtHints : tactic

def parseHints : TSyntax `smtHints → TacticM (List Expr)
  | `(smtHints| [ $[$hs],* ]) => hs.toList.mapM (fun h => elabTerm h.raw none)
  | `(smtHints| ) => return []
  | _ => throwUnsupportedSyntax

def parseTimeout : TSyntax `smtTimeout → TacticM (Option Nat)
  | `(smtTimeout| (timeout := $n)) => return some n.getNat
  | `(smtTimeout| ) => return some 5
  | _ => throwUnsupportedSyntax

def prepareSmtQuery (hints : TSyntax `smtHints) : TacticM (List Command) := do
  -- 1. Get the current main goal.
  let goalType ← Tactic.getMainTarget
  let goalId ← Lean.mkFreshMVarId
  Lean.Meta.withLocalDeclD goalId.name (mkNot goalType) fun g => do
  -- 2. Get the hints passed to the tactic.
  let mut hs ← parseHints hints
  hs := hs.eraseDups
  -- 3. Generate the SMT query.
  Query.generateQuery g hs

def elabProof (text : String) : TacticM Unit := do
  let (env, log) ← process text (← getEnv) .empty "<proof>"
  _ ← modifyEnv (fun _ => env)
  for m in log.msgs do
    trace[smt.debug.reconstruct] (← m.toString)
  if log.hasErrors then
    throwError "encountered errors elaborating cvc5 proof"

def evalAnyGoals (tactic : TacticM Unit) : TacticM Unit := do
  let mvarIds ← getGoals
  let mut mvarIdsNew := #[]
  for mvarId in mvarIds do
    unless (← mvarId.isAssigned) do
      setGoals [mvarId]
      try
        tactic
        mvarIdsNew := mvarIdsNew ++ (← getUnsolvedGoals)
      catch _ =>
        mvarIdsNew := mvarIdsNew.push mvarId
  setGoals mvarIdsNew.toList

private def addDeclToUnfoldOrTheorem (thms : Meta.SimpTheorems) (e : Expr) : MetaM Meta.SimpTheorems := do
  if e.isConst then
    let declName := e.constName!
    let info ← getConstInfo declName
    if (← Meta.isProp info.type) then
      thms.addConst declName
    else
      thms.addDeclToUnfold declName
  else
    thms.add (.fvar e.fvarId!) #[] e

open Reconstruction.Certifying in
def rconsProof (hints : List Expr) : TacticM Unit := do
  let mut gs ← (← Tactic.getMainGoal).apply (mkApp (mkConst ``notNotElim) (← Tactic.getMainTarget))
  Tactic.replaceMainGoal gs
  let th0 ← Meta.mkConstWithFreshMVarLevels `th0
  trace[smt.debug.reconstruct] "th0 : {← Meta.inferType th0}"
  gs ← (← Tactic.getMainGoal).apply th0
  Tactic.replaceMainGoal gs
  for h in hints do
    evalAnyGoals do
      let gs ← (← Tactic.getMainGoal).apply h
      Tactic.replaceMainGoal gs
  let mut some thms ← (← Meta.getSimpExtension? `smt_simp).mapM (·.getTheorems)
    | throwError "smt tactic failed, 'smt_simp' simpset is not available"
  -- TODO: replace with our abbreviation of `Implies`
  thms ← thms.addDeclToUnfold ``Implies
  for h in hints do
    thms ← addDeclToUnfoldOrTheorem thms h
  evalAnyGoals do
    let (result?, _) ← Meta.simpGoal (← Tactic.getMainGoal) {
      simpTheorems := #[thms],
      congrTheorems := (← Meta.getSimpCongrTheorems)
    }
    match result? with
    | none => replaceMainGoal []
    | some (_, mvarId) => replaceMainGoal [mvarId]

@[tactic smt] def evalSmt : Tactic := fun stx => withMainContext do
  let goalType ← Tactic.getMainTarget
  let cmds ← prepareSmtQuery ⟨stx[1]⟩
  let query := setOption "produce-models" "true"
            *> emitCommands cmds.reverse
            *> checkSat
  logInfo m!"goal: {goalType}"
  logInfo m!"\nquery:\n{Command.cmdsAsQuery (.checkSat :: cmds)}"
  -- 4. Run the solver.
  let kind := smt.solver.kind.get (← getOptions)
  let path := smt.solver.path.get? (← getOptions)
  let timeout ← parseTimeout ⟨stx[2]⟩
  let ss ← createFromKind kind path timeout
  let (res, ss) ← (StateT.run query ss : MetaM _)
  -- 5. Print the result.
  logInfo m!"\nresult: {res}"
  if res = .sat then
    -- 5a. Print model.
    let (model, _) ← StateT.run getModel ss
    logInfo m!"\ncounter-model:\n{model}\n"
    throwError "unable to prove goal, either it is false or you need to define more symbols with `smt [foo, bar]`"
  if res = .unknown then
    throwError "unable to prove goal"
  try
    -- 5a. Reconstruct proof.
    let (.expr [.atom "proof", .atom nnp], _) ← StateT.run getProof ss
      | throwError "encountered error parsing cvc5 proof"
    let nnp := skipImports (unquote nnp)
    trace[smt.debug.reconstruct] "proof:\n{nnp}"
    elabProof nnp
    rconsProof (← parseHints ⟨stx[1]⟩)
  catch e =>
    logInfo m!"failed to reconstruct proof: {e.toMessageData}"
where
  unquote s := s.extract ⟨1⟩ (s.endPos - ⟨1⟩)
  skipImports (s : String) := Id.run do
    let mut s := s
    while s.startsWith "import" do
      s := s.dropWhile (· != '\n')
      s := s.drop 1
    return s

@[tactic smtShow] def evalSmtShow : Tactic := fun stx => withMainContext do
  let goalType ← Tactic.getMainTarget
  let cmds := .checkSat :: (← prepareSmtQuery ⟨stx[1]⟩)
  -- 4. Print the query.
  logInfo m!"goal: {goalType}\n\nquery:\n{Command.cmdsAsQuery cmds}"

end Smt
