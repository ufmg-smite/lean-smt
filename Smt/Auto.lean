/-
Copyright (c) 2021-2024 by the authors listed in the file AUTHORS and their
institutional affiliations. All rights reserved.
Released under Apache 2.0 license as described in the file LICENSE.
Authors: Abdalrhman Mohamed
-/

import Auto.Tactic
import Smt.Tactic.Smt

open Lean in
def smtSolverFunc (ls : Array Auto.Lemma) (is : Array Auto.Lemma) : MetaM Expr := do
  let fi l := do
    let userName ← mkFreshUserName `inst
    let type ← Meta.mkAppM ``Inhabited #[l.type]
    let value ← Meta.mkAppOptM ``Inhabited.mk #[l.type, l.proof]
    return { userName, type, value }
  let is ← is.mapM fi
  let fl l := do
    let userName ← mkFreshUserName `h
    let type ← Auto.Lam2D.approxSimpNF l.type
    let value := l.proof
    return { userName, type, value }
  let ls ← ls.mapM fl
  let h ← Meta.mkFreshExprMVar (Expr.const ``False [])
  let (_, mv) ← h.mvarId!.assertHypotheses is
  let (fvs, mv) ← mv.assertHypotheses ls
  mv.withContext do
    let hs := fvs.map (.fvar ·)
    _ ← Smt.smt mv hs.toList
    -- Note: auto should allow solvers to export new goals to users
    -- for mv in mvs do
    --   logInfo m!"new : {mv}"
    instantiateMVars h

attribute [rebind Auto.Native.solverFunc] smtSolverFunc
