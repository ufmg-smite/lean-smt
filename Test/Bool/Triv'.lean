import Smt

theorem triv': ∀ p : Bool, ∀ _ : p, p := by
  smt
  intro p
  simp_all
