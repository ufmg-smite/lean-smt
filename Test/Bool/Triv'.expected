goal: ∀ (p : Bool), p = true → p = true

query:
(assert (not (forall ((p Bool)) (=> (= p true) (= p true)))))
(check-sat)

result: unsat
failed to reconstruct proof: tactic 'apply' failed, failed to unify
  (¬∀ (p : Prop), Implies (p = True) (p = True)) → False
with
  ¬¬∀ (p : Bool), p = true → p = true
⊢ ¬¬∀ (p : Bool), p = true → p = true
